src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,562 @@
+					state(i, j, k, HydroSystem<NewProblem>::energy_index) += rho_eint_blast;
+					state(i, j, k, HydroSystem<NewProblem>::internalEnergy_index) += rho_eint_blast;
+					state(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex) += scalar_blast;
+
+					printf("The total number of SN gone off=%d\n", cum_sn);
+					Real Rpds = 14. * std::pow(state(i, j, k, HydroSystem<NewProblem>::density_index) / Const_mH, -3. / 7.);
+					printf("Rpds = %.2e pc\n", Rpds);
+				}
+			}
+		});
+	}
+}
+
+template <> void QuokkaSimulation<NewProblem>::computeBeforeTimestep()
+{
+	// compute how many (and where) SNe will go off on the this coarse timestep
+	// sample from Poisson distribution
+
+	const Real dt_coarse = dt_[0];
+	const Real domain_area = geom[0].ProbLength(0) * geom[0].ProbLength(1);
+	const Real mean = 0.0;
+	const Real stddev = hscale / geom[0].ProbLength(2) / 2.;
+
+	const Real expectation_value = ks_sigma_sfr * domain_area * dt_coarse;
+
+	const int count = static_cast<int>(amrex::RandomPoisson(expectation_value));
+
+	// resize particle arrays
+	amrex::Array<int, 1> const lo{0};
+	amrex::Array<int, 1> const hi{count};
+	userData_.blast_x = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.blast_y = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.blast_z = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.nblast = count;
+	userData_.SN_counter_cumulative += count;
+
+	// for each, sample location at random
+	auto const &px = userData_.blast_x->table();
+	auto const &py = userData_.blast_y->table();
+	auto const &pz = userData_.blast_z->table();
+	for (int i = 0; i < count; ++i) {
+		px(i) = geom[0].ProbLength(0) * amrex::Random();
+		py(i) = geom[0].ProbLength(1) * amrex::Random();
+		pz(i) = geom[0].ProbLength(2) * amrex::RandomNormal(mean, stddev);
+	}
+}
+
+template <> void QuokkaSimulation<NewProblem>::computeAfterLevelAdvance(int lev, amrex::Real time, amrex::Real dt_lev, int ncycle)
+{
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = geom[lev].ProbLoArray();
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_hi = geom[lev].ProbHiArray();
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dx = geom[lev].CellSizeArray();
+
+	AddSupernova(state_new_cc_[lev], prob_lo, prob_hi, dx, userData_, lev);
+}
+
+template <>
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE auto
+HydroSystem<NewProblem>::GetGradFixedPotential(amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> posvec) -> amrex::GpuArray<amrex::Real, AMREX_SPACEDIM>
+{
+
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> grad_potential;
+	grad_potential[0] = 0.0;
+	grad_potential[1] = 0.0;
+
+	double z = posvec[2];
+
+	// Interpolate to find the accurate g-value from array-- because linterp doesn't work on Setonix
+	size_t i = 0;
+	double x_interp = std::abs(z);
+	amrex::GpuArray<amrex::Real, 100> xx = z_data;
+	amrex::GpuArray<amrex::Real, 100> yy = logg_data;
```

Comment:
This also makes a copy of a large array, which is not a good idea. You can change this to a reference, or you use `z_data` and `logg_data` directly.